schemas/devContainerTemplate.schema.json

```diff
@@ -0,0 +1,182 @@
+          "description": "The JSON schema of the `devcontainer-template.json` file.",
+          "type": "string",
+          "format": "uri"
+        },
+        "description": {
+          "description": "Description of the Template.",
+          "type": "string"
+        },
+        "documentationURL": {
+          "description": "Url that points to the documentation of the Template.",
+          "type": "string",
+          "format": "uri"
+        },
+        "id": {
+          "description": "ID of the Template. The id should be unique in the context of the repository/published package where the Template exists and must match the name of the directory where the devcontainer-template.json resides.",
+          "type": "string"
+        },
+        "keywords": {
+          "description": "List of strings relevant to a user that would search for this Template.",
+          "type": "array",
+          "items": {
+            "type": "string"
+          }
+        },
+        "licenseURL": {
+          "description": "Url that points to the license of the Template.",
+          "type": "string",
+          "format": "uri"
+        },
+        "name": {
+          "description": "Name of the Template.",
+          "type": "string"
+        },
+        "options": {
+          "description": "A map of options that the supporting tools should use to populate different configuration options for the Template.",
+          "type": "object",
+          "additionalProperties": {
+            "$ref": "#/definitions/Option"
+          }
+        },
+        "platforms": {
+          "description": "Languages and platforms supported by the Template.",
+          "type": "array",
+          "items": {
+            "type": "string"
+          }
+        },
+        "publisher": {
+          "description": "Name of the publisher/maintainer of the Template.",
+          "type": "string"
+        },
+        "version": {
+          "description": "The semantic version of the Template.",
+          "type": "string"
+        }
+      },
+      "required": [
+        "id",
+        "name",
+        "version"
+      ]
+    },
+    "Option": {
+      "description": "An option that the supporting tools should use to populate different configuration options for the Template.",
+      "type": "object",
+      "properties": {
+        "default": {
+          "description": "Default value for the option.",
+          "type": [
+            "boolean",
+            "string"
+          ]
+        },
+        "description": {
+          "description": "Description for the option.",
+          "type": "string"
+        },
+        "type": {
+          "description": "Type of the option. Valid types are currently: boolean, string",
+          "type": "string",
+          "enum": [
+            "boolean",
+            "string"
+          ]
+        }
+      },
+      "required": [
+        "default",
+        "description",
+        "type"
+      ],
+      "oneOf": [
```

Comment:
devContainerFeature.schema.json uses `anyOf` here and has a few more descriptions filled in. Maybe copy from that?

Reply:
I'll update the `oneOf` to `anyOf` to match the Feature schema and do another pass on the descriptions; I didn't think I'd missed any. 🙁

Reply:
Okay, I *completely* misunderstood what you were referring to here. 😅

I used `oneOf` here because it more clearly communicates that *exactly one* of the subschemas (boolean option, string option) must apply. In this case, they're incompatible (`"const": "boolean"` vs. `"const": "string"`), but even if they weren't, it wouldn't be reasonable for an option to be considered both a boolean and a string.

Similarly, the placement of the descriptions are deliberate, to avoid redundancy. The descriptions for the properties which are present regardless of the option's type (`default`, `description`, and `type`) are present at the top level so that those properties don't need a subschema with any more detail than is particular to the `oneOf` variant selected; you can see this most easily with the `description` property, which has the subschema of just `true` for both boolean options and string options.

Admittedly, I've only tested this with VS Code, but it behaves as I would expect it to: the descriptions for all properties are available regardless of which `oneOf` variant has been matched, or even if neither has yet matched.

!["default" property description (defined in base option schema), no option type](https://github.com/devcontainers/spec/assets/275142/ab0e7f57-2eee-4109-860f-51d75ae0bdfb)
*"default" property description (defined in base option schema), no option type*

!["default" property description (defined in base option schema), option has "string" type](https://github.com/devcontainers/spec/assets/275142/da8c28bf-8582-49c1-a8c5-bc51de192058)
*"default" property description (defined in base option schema), option has "string" type*

!["proposals" property description (defined in string option subschema), no option type](https://github.com/devcontainers/spec/assets/275142/6a214de7-9dee-466a-85a8-16c164a9396d)
*"proposals" property description (defined in string option subschema), no option type*

!["proposals" property description (defined in string option subschema), option has "string" type](https://github.com/devcontainers/spec/assets/275142/beaff862-8dcc-4be1-9711-144a6e30ffbb)
*"proposals" property description (defined in string option subschema), option has "string" type*

Reply:
I see, it should be `oneOf` in the features schema then, could you update it there? Thanks!

Replicating the descriptions might still make sense for other tools and helps when someone reads the schema themselves. 🤔

Reply:
No worries, I've made that change to the feature schema. While I was in there, I noticed that "proposals" isn't required when "enum" is absent and have made the same change to the template schema. (Thinking back, I don't know why I made it required in the first place.)

I'm a bit torn over the descriptions as well. On the one hand, schemas are primarily intended for machine consumption rather than being read directly by humans… but on the other hand, a lot of us read them anyway! As far as other consumers go, my experience with schema tools other than VS Code is pretty much limited to validation libraries, which I don't believe I've seen use descriptions from the schema.

Ultimately, I came down on the side of "redundancy is bad" simply because future updates can't cause duplicate descriptions to drift out of sync if there aren't duplicate descriptions in the first place. If there's a stronger argument for duplication that just hasn't occurred to me, I'm happy to make the change!